### PR TITLE
calculateSharesDeltaGivenBondsDeltaDerivativeSafe edge case

### PR DESCRIPTION
**Pull request by mcclurejt**
# Resolved Issues

fixes https://github.com/spearbit-audits/delv-review/issues/18

# Description

Returns an error when `rhs == ONE` however it can successfully return 0 here.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed. If there are
multiple reviewers, copy the checklists into sections titled `## [Reviewer Name]`.
If the PR doesn't touch Solidity and/or Rust, the corresponding checklist can
be removed.

### Solidity

- [x] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [x] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [x] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [x] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [x] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [x] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [x] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### Solidity

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### CODE REVIEW

**Review by jalextowle — APPROVED** [review]
Approving this since the fix is obviously correct and testing this seems really gnarly. If you can figure it out, I'd be pretty stoked.

**Comment by mcclurejt** [comment]
> Approving this since the fix is obviously correct and testing this seems really gnarly. If you can figure it out, I'd be pretty stoked.

Flagging that during initial testing with setting `shareAdjustment = shareReserves`, the function actually reverts despite being labeled as `Safe`. AFAIK this situation could never occur in the real world, but figured I'd mention in case I'm wrong

**Comment by jalextowle** [comment]
> > Approving this since the fix is obviously correct and testing this seems really gnarly. If you can figure it out, I'd be pretty stoked.
> 
> @jalextowle
> 
> Flagging that during initial testing with setting `shareAdjustment = shareReserves`, the function actually reverts despite being labeled as `Safe`. After investigating, the cause is [this block here](https://github.com/delvtech/hyperdrive/blob/1c553a0f9d971f41990e11176d718997e4705274/contracts/src/libraries/LPMath.sol#L1234:L1247) (assuming the culprit is a divide by zero), however there could be more. I've committed my test case so you can have a look if you choose...
> 
> The case of `shareAdjustment = shareReserves` is also required for us to test this fix, since it would result in `rhs == ONE` in this context.
> 
> Is this scenario possible in a real world environment? Even if not, would it be worth it to include checks that avoid the revert in line with the function's `Safe` designation?
> 
> Will be reviewing the whitepaper/code to see if I can get an answer on my own as well.

This shouldn't ever happen due to the restrictions that we place on the `effectiveShareReserves` in the long and short flows and because we burn some LP shares during `initialize` (see this test for more details about how this impacts the effective share reserves: https://github.com/delvtech/hyperdrive/blob/f2a578765ec58f0c1dc5235f743da5f53a51ec74/test/integrations/hyperdrive/LPWithdrawalTest.t.sol#L1917). If you find a case that violates this assumption, that would be a critical vulnerability, so it's definitely worth putting some fresh eyes on it!

File: test/units/libraries/LPMath.t.sol [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { MockLPMath } from "contracts/test/MockLPMath.sol";
 import { HyperdriveUtils } from "test/utils/HyperdriveUtils.sol";
 import { HyperdriveTest } from "test/utils/HyperdriveTest.sol";
 import { Lib } from "test/utils/Lib.sol";
+import "forge-std/console2.sol";
```
[the remark this answers is not in the document]
Comment (mcclurejt) [reply test/units/libraries/LPMath.t.sol]: ty!

**Pull request merged**